### ISSUE TITLE
Proposal: add the ability of distributed Tracing

### ISSUE THREAD

**Issue by LinuxSuRen** [argoproj/argo-workflows#12077]
# Summary

Currently, we could leverage Prometheus to see some important metrics of Argo Workflow. But it would be wonderful if we could see the tracing, so it's helpful for checking potential neckbottle.

## Use Cases

[Apache Skywalking](https://skywalking.apache.org/) is a very popular tool which have both front-end and backend for the tracing feature. And the news is that we don't need modify any current code lines to archive that. [Apache Skywalking Go Agent](https://github.com/apache/skywalking-go) will do it during the compile process.

I'd be happy to do it if I could got positive feddback.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

**Comment by terrytangyuan** [comment]
Can users use it directly if it doesn't require any modifications in Argo Workflows? What's left to be done in Argo?

**Comment by LinuxSuRen** [comment]
Users need to provide the endpoint of SkyWalking during the deployment process once Argo Workflow adds the related build flags like `go build -toolexec="skywalking-go-agent"`.

It can be traced in a variety of ways. For now, it includes gRPC, HTTP, etc.

See an example https://github.com/LinuxSuRen/api-testing/blob/7228009f26d0ac5fdef19a25600e29e2610a74d6/Dockerfile#L40

So, below is the list of people need to do:

* User, add the related environment variables
* Dev, modify the Dockerfile or Makefile

**Comment by terrytangyuan** [comment]
I feel like that type of instructions does not have to be in this repo.

**Comment by LinuxSuRen** [comment]
I didn't get your point. Could you be more specific?

**Comment by terrytangyuan** [comment]
> So, below is the list of people need to do

That list includes the things users need to do, right? Then it should be live in SkyWalking's repo instead of argo-workflows repo. We cannot add flags like `"skywalking-go-agent"` in this repo.

**Comment by LinuxSuRen** [comment]
`skywalking-go-agent` is a Golang build tool plugin, it could inject some intercepter code into the target project. So, the target project don't need to change the business code lines. Let me try to make it be more clear.

To end users, the could specific the endpoint of SkyWalking server which will collect the tracing data.

To the developer of Argo workflows, they need to update the compiling process. AKA modify the Makefile and Dockerfile.

**Comment by agilgur5** [comment]
Optional distributed tracing support could potentially be good to have. It could be enabled by an environment variable and the collector URL and port can be specified in an environment variable as well. We'd want to add it to both the Controller and Server.

Re: Apache Skywalking, is there a reason you prefer that over the standard [OTel Go library](https://github.com/open-telemetry/opentelemetry-go)?

> Can users use it directly if it doesn't require any modifications in Argo Workflows? What's left to be done in Argo?

@terrytangyuan for distributed tracing, some library is needed to instrument the code and make it output traces. We should be vendor neutral, which is why the standard OTel libraries would make sense to me (though I would think Apache Skywalking is also vendor neutral).
Similar to how Prometheus requires an application to have an Exporter for metrics.

> so it's helpful for checking potential neckbottle.

For single application tracing, Argo does already support `pprof` via optional [env var](https://argoproj.github.io/argo-workflows/environment-variables/#controller).
Distributed tracing primarily helps with requests that span across multiple micro-services. Although it is also useful just to connect to the OTel tooling ecosystem

**Comment by LinuxSuRen** [comment]
> Apache Skywalking, is there a reason you prefer that over the standard [OTel Go library](https://github.com/open-telemetry/opentelemetry-go)?

OpenTelemetry is great. However, it requires users to write the instrument code in Argo workflows. SkyWalking did that work automatically, and inject the code into Argo workflow. So, for the second one, it could keep Argo workflow's code be clean.

**Comment by LinuxSuRen** [comment]
I tried it locally. Share some screenshots here.

![image](https://github.com/argoproj/argo-workflows/assets/1450685/0a6d5821-66fa-4104-af7b-9e2f371c04f5)
![image](https://github.com/argoproj/argo-workflows/assets/1450685/1f3871de-3c48-44aa-9006-6026c151cc7b)
![image](https://github.com/argoproj/argo-workflows/assets/1450685/79b78128-2f77-4f5e-833d-4095b4dd145c)

**Comment by agilgur5** [comment]
> OpenTelemetry is great. However, it requires users to write the instrument code in Argo workflows. SkyWalking did that work automatically, and inject the code into Argo workflow. So, for the second one, it could keep Argo workflow's code be clean.

I took a look at the code for the Go Agent ([gRPC](https://github.com/apache/skywalking-go/blob/c4d27b394d31f2313ff7d3078ca75d635e6a1565/plugins/grpc/instrument.go) for example) and I can see it does an AST walk to add interceptors. Although I see there are still are [manual trace APIs](https://skywalking.apache.org/docs/skywalking-go/next/en/advanced-features/manual-apis/toolkit-trace/), so there can still be code in Argo's codebase

I also checked to see if it can be configured at run-time via env vars (e.g. for connecting to collectors) and it seems like [it can be](https://skywalking.apache.org/docs/skywalking-go/next/en/advanced-features/settings-override/#environment-variables).
Although it's not clear to me if it can be entirely disabled at run-time from [the configuration](https://github.com/apache/skywalking-go/tree/c4d27b394d31f2313ff7d3078ca75d635e6a1565/tools/go-agent/config/agent.default.yaml) -- there's no `enabled: true` config (as is common in Helm, for instance).
If we were to instrument ourselves, we would not add instrumentation what-so-ever if tracing is not enabled via env var.

> I tried it locally. Share some screenshots here.

Thanks for testing, was going to ask for some example usage.

Do you know if the Go agent can directly connect to OTel collectors? I would like to see if OTel tooling will work out-of-the-box.

It seems like Skywalking has its own trace format, as the Skywalking server has [OTel](https://skywalking.apache.org/docs/main/next/en/setup/backend/otlp-trace/) and [Zipkin](https://skywalking.apache.org/docs/main/next/en/setup/backend/zipkin-trace) converters for it to be able to interpret those formats. 

For instance, it seems like Grafana Tempo and Jaeger [both](https://github.com/grafana/tempo/issues/466) [don't support](https://github.com/jaegertracing/jaeger/issues/1669) Skywalking.
The [OTel Collector](https://github.com/open-telemetry/opentelemetry-collector) doesn't seem to support it in core either, though it looks like there are contrib components for a [Skywalking Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/c50daa9/receiver/skywalkingreceiver) and a [Skywalking Exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/c50daa9/exporter/skywalkingexporter)

**Comment by LinuxSuRen** [comment]
Thanks for the detailed reply. Users can choose if send the tracing data to SkyWalking server [via env](https://github.com/apache/skywalking-go/pull/134). At the meaning time, I'm still exploring OpenTelementry as well. The SK agent sending the data via SkyWalking gRPC interface, it seems do not support send data to OpenTelementry for now.

**Comment by agilgur5** [comment]
> Users can choose if send the tracing data to SkyWalking server [via env](https://github.com/apache/skywalking-go/pull/134).

Oh cool, I saw that the reporter had a [`discard`](https://github.com/apache/skywalking-go/blob/c4d27b394d31f2313ff7d3078ca75d635e6a1565/tools/go-agent/config/agent.default.yaml#L35) config, no idea that you were the one that added it very recently!

I was thinking of going further than that though, don't create spans _at all_ if it's not enabled. Then there's no CPU or memory overhead if you're not using distributed tracing.

**Comment by LinuxSuRen** [comment]
> don't create spans / run instrumentation at all if it's not enabled

I totally agree with you. Discarding the data is an easy solution. I'll try to find a way to disable it eventually. But no guraranteed time could be given.

**Comment by agilgur5** [comment]
Saw [this thread in the `#argo-contributors`](https://cloud-native.slack.com/archives/C020XM04CUW/p1699825880000639) Slack channel today and Argo CD apparently implemented OTel already. https://github.com/argoproj/argo-cd/pull/7539 was linked there, but that seems to be for gRPC specifically and it looks like there might be other code there. We'd definitely want to stay as consistent as possible between Argo projects

**Comment by ouyuran** [comment]
Is this proposal still valid? I instrumented our project before with traces/logs/metrics, and now we are considering migrate to Argo Workflow and I found this proposal.
I guess I can either start from a sidecar, hooks and instrumented application docker images, or do it in Argo Workflow itself.
Back to the original question, are you still looking for someone who can contribute to this?

**Comment by certara-gschaffer** [comment]
@Joibel according to this talk you guys have a tracing solution planned for 3.7 originally. What happend to it?

https://www.youtube.com/watch?v=F0-D99NqBjs
